cmd/uzfs/libuzfs_server.c

```diff
@@ -0,0 +1,295 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License (the "License").
+ * You may not use this file except in compliance with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+#include <arpa/inet.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <libuzfs.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <pthread.h>
+#include <signal.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <sys/un.h>
+#include <sys/zfs_context.h>
+#include <unistd.h>
+
+void close_client(int sock, void *arg);
+
+/*
+ * Block SIGPIPE signal
+ */
+static int
+uzfs_server_init(void)
+{
+	sigset_t set;
+	if (sigemptyset(&set) < 0)
+		return (-1);
+	if (sigaddset(&set, SIGPIPE) < 0)
+		return (-1);
+	if (pthread_sigmask(SIG_BLOCK, &set, NULL))
+		return (-1);
+	return (0);
+}
+
+/*
+ * copy sock fd to *arg and close it
+ */
+void
+close_client(int sock, void *arg)
+{
+	int64_t fd = (int64_t)arg;
+	/*
+	 * have to close the fd and also have to make sure this
+	 * fd does not get assigned. Both should be happening atomically.
+	 */
+	if (dup2(sock, fd) < 0) {
+		perror("close_client dup2");
+	}
+}
+
+/*
+ * monitor the given socket and execute the relevant action
+ */
+static void *
+uzfs_monitor_socket(void *arg)
+{
+	uzfs_mon_t *mon = arg;
+	int sock = mon->mon_fd;
+	int count = 0;
+
+	while (1) {
+		char c;
+		int x = recv(sock, &c, sizeof (c), MSG_PEEK);
+		if (x == 0) {
+			mon->mon_action(sock, mon->mon_arg);
+			break;
+		}
+		count++;
+		sleep(1);
+	}
+
+	pthread_exit(NULL);
+}
+
+/*
+ * once control reaches to our uZFS process, it does not know anything
+ * about the client. So if we did ctrl + c or killed the zfs process,
```

Comment:
I think the default should be to mimic the behavior with the kernel module.
That is, the client `zfs` process cannot be killed while waiting for a response from the `uZFS` daemon.

Of course, we cannot prohibit `SIGKILL` for the client `zfs` process, but we can mask `SIGINT` and such.

Reply:
I think blocking `SIGINT` may not be the right behavior since the user may use the other signal to kill the process. So the right way would be to handle this in `uZFS` daemon.

Reply:
Sure, they might be able to kill the process through other signals.
However, users are currently accustomed to Ctrl+C not succeeding until the ioctl returns and the ZFS operation is complete or cancelled.
Unless the operation triggers some background activity (like zpool scrub), the general assumption is that the `zfs` management operations are done or cancelled once the `zfs` process exits.
I believe that failing to uphold that assumption will lead to lots of headaches, e.g., `EBUSY` or similar when Ctrl+C'ing and immediately re-running the same command.
If we go with @ahrens 's proposal to `using an environment variable or CLI flag to determine whether /sbin/zfs talks to the uZFS process vs. the kernel` then the `zfs` binary should behave as similar as possible in both configurations.

Reply:
Alternatively / additionally, for debugging, some subcommand to view ongoing activity would be useful. E.g. `zfs uzfs ps` or similar to show the threads, and maybe the client's PID as well (`SO_PEERCRED`, see https://stackoverflow.com/questions/8104904/identify-program-that-connects-to-a-unix-domain-socket )